src/main/kotlin/com/exactpro/th2/processor/core/state/DataProviderStateStorage.kt

```diff
@@ -0,0 +1,339 @@
+
+@NotThreadSafe
+class DataProviderStateStorage(
+    private val messageRouter: MessageRouter<MessageGroupBatch>,
+    private val eventBatcher: EventBatcher,
+    private val dataProvider: DataProviderService,
+    private val bookName: String,
+    private val stateSessionAlias: String,
+    maxMessageSize: Long = METADATA_SIZE + MIN_STATE_SIZE
+) : IStateStorage {
+
+    init {
+        check(maxMessageSize - METADATA_SIZE >= MIN_STATE_SIZE) {
+            "`Max message size` too small, because max message size ($maxMessageSize) - metadata size ($PROPERTY_SIZE) < min state size $MIN_STATE_SIZE"
+        }
+    }
+
+    private val stateSize = maxMessageSize - METADATA_SIZE
+    private var sequenceCounter: Long = System.nanoTime()
+
+    private val requestBuilder = MessageSearchRequest.newBuilder().apply {
+        searchDirection = PREVIOUS
+        resultCountLimit = Int32Value.of(COUNT_LIMIT)
+        addResponseFormats(RAW_MESSAGE_RESPONSE_FORMAT)
+        bookIdBuilder.apply {
+            name = bookName
+        }
+        addStreamBuilder().apply {
+            name = stateSessionAlias
+            direction = Direction.SECOND
+        }
+    }
+
+    override fun loadState(parentEventId: EventID): ByteArray? {
+        K_LOGGER.info { "Started state loading" }
+
+        val state = mutableListOf<Pair<StateType, MessageGroupResponse>>()
+        sequence {
+            var iterator = dataProvider.searchMessages(createSearchRequest().also {
+                K_LOGGER.info { "Request to load state ${it.toJson()}" }
+            })
+
+            while (iterator.hasNext()) {
+                val message: MessageGroupResponse = iterator.next().message
+                yield(message)
+                if (!iterator.hasNext()) {
+                    iterator = dataProvider.searchMessages(createSearchRequest(message.messageId.timestamp).also {
+                        K_LOGGER.info { "Request to load state ${it.toJson()}" }
+                    })
+                }
+            }
+        }.map(::mapStateTypeToMessage)
+        .dropWhile { (stateType, _) -> !(stateType == SINGLE || stateType == END) }
+        .forEach { pair ->
+            val (stateType, response) = pair
+            K_LOGGER.debug { "Process raw response ${response.toJson()}" }
+            when (stateType) {
+                SINGLE -> {
+                    reportStateLoaded(parentEventId, pair)
+                    return response.bodyRaw.toByteArray()
+                }
+                END, INTERMEDIATE, START -> {
+                    if (state.checkAndModify(pair)) {
+                        val list: List<Byte> =
+                            state.reversed().flatMap { (_, message) -> message.bodyRaw.toByteArray().asSequence() }
+                        reportStateLoaded(parentEventId, state)
+                        return ByteArray(list.size).apply { list.forEachIndexed(this::set) }
+                    }
+                }
+                else -> {
```

Comment:
It is better to use the UNKNOWN value here. It will allow us to make the when statement exhaustive. And will have a compilation error if (for some reason) we add a new enum value